# **32147009 - Maryam Saeed- Databases and Analytics**

# **MongoDB Development**

**Installing and Importing Required Libraries**

The required Python libraries were installed and imported to support MongoDB connection, data handling, and preprocessing tasks. PyMongo was used for database operations, while Pandas was used for data manipulation and analysis in a structured format.

In [1]:
# Install & Import
!pip install pymongo
from pymongo import MongoClient
from pymongo import ASCENDING, DESCENDING
import pandas as pd
import json

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.8/1.8 MB 56.7 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 331.1/331.1 kB 22.1 MB/s eta 0:00:00


**Uploading Dataset Files**

All dataset files were uploaded into the working environment using Google Colab. These files contain operational data related to customers, drivers, vehicles, orders, deliveries, complaints, incidents, hubs, and app events.

In [2]:
from google.colab import files
uploaded = files.upload()

Saving app_events.csv to app_events.csv
Saving complaints.csv to complaints.csv
Saving customers.csv to customers.csv
Saving deliveries.csv to deliveries.csv
Saving drivers.csv to drivers.csv
Saving hubs.csv to hubs.csv
Saving incidents.csv to incidents.csv
Saving orders.csv to orders.csv
Saving vehicles.csv to vehicles.csv


**Loading CSV Files into DataFrames**

All uploaded CSV files were loaded into separate Pandas dataframes. Each dataframe represents a different entity in the NorthStar Urban Mobility system, allowing structured processing and preparation for MongoDB insertion.

In [3]:
# Load all the files
app_events  = pd.read_csv('/content/app_events.csv')
complaints  = pd.read_csv('/content/complaints.csv')
customers   = pd.read_csv('/content/customers.csv')
deliveries  = pd.read_csv('/content/deliveries.csv')
drivers     = pd.read_csv('/content/drivers.csv')
hubs        = pd.read_csv('/content/hubs.csv')
incidents   = pd.read_csv('/content/incidents.csv')
orders      = pd.read_csv('/content/orders.csv')
vehicles    = pd.read_csv('/content/vehicles.csv')

print("✅ All files loaded")

✅ All files loaded


**Filling Missing Values**

Missing values were identified and handled across multiple datasets to improve data quality and ensure accurate analysis. Different techniques such as replacing missing values with median, mode, zero values, or default labels were applied depending on the data type and business meaning of each attribute.

In [4]:
app_events['order_id'] = app_events['order_id'].fillna("NoOrder")
complaints['compensation_amount'] = complaints['compensation_amount'].fillna(0)
customers['loyalty_score'] = customers['loyalty_score'].fillna(customers['loyalty_score'].median())
customers['preferred_channel'] = customers['preferred_channel'].fillna(customers['preferred_channel'].mode()[0])
deliveries['customer_rating_post_delivery'] = deliveries['customer_rating_post_delivery'].fillna(0)
drivers['training_score'] = drivers['training_score'].fillna(drivers['training_score'].median())
incidents['resolved_hours'] = incidents['resolved_hours'].fillna(incidents['resolved_hours'].median())
orders['booking_channel'] = orders['booking_channel'].fillna(orders['booking_channel'].mode()[0])
vehicles['battery_health_pct'] = vehicles['battery_health_pct'].fillna(vehicles['battery_health_pct'].median())

print("✅ Missing values filled")

✅ Missing values filled


**Fixing Zone Inconsistencies**

Zone inconsistencies were corrected by standardising zone names across all datasets. Different representations of the same geographic zones were converted into a consistent format to avoid errors during grouping, filtering, and analysis operations.

In [5]:
zone_map = {
    'north':'North','NORTH':'North','SOUTH':'South','south':'South',
    'EAST':'East','east':'East','WEST':'West','west':'West',
    'CENTRAL':'Central','central':'Central','Ctr':'Central',
    'AIRPORT':'Airport','airport':'Airport',
    'Riverside':'RiverSide','RIVERSIDE':'RiverSide',
}
def fix_zones(df, cols):
    for col in cols:
        if col in df.columns:
            df[col] = df[col].map(lambda x: zone_map.get(x, x) if pd.notna(x) else x)
    return df

customers  = fix_zones(customers,  ['home_zone'])
drivers    = fix_zones(drivers,    ['base_zone'])
vehicles   = fix_zones(vehicles,   ['assigned_zone'])
orders     = fix_zones(orders,     ['pickup_zone','dropoff_zone'])
app_events = fix_zones(app_events, ['zone_context'])

print("✅ Fixed zone inconsistencies ")

✅ Fixed zone inconsistencies 


**Connecting to MongoDB Atlas**

A connection was established to MongoDB Atlas using PyMongo. The database named northstar_db was accessed to store and manage all datasets. This enables cloud-based storage and efficient handling of large-scale operational data.

In [6]:
Client = MongoClient("mongodb+srv://maryamsaeed_db:Maryamdb28@cluster0.80ofzwx.mongodb.net/?appName=Cluster0")
db = Client["northstar_db"]
print("✅ Connected to MongoDB Atlas")

✅ Connected to MongoDB Atlas


**Referenced data examples**

In [ ]:
db.orders.find_one()

{'_id': ObjectId('6a02e1fdb9800d3fb76c0f28'),
 'order_id': 'O00001',
 'customer_id': 'C0292',
 'service_type': 'Passenger',
 'order_created_at': '2024-08-20 14:43:00',
 'promised_window_hours': 6,
 'pickup_zone': 'Airport',
 'dropoff_zone': 'South',
 'priority_level': 'Medium',
 'order_value': 126.65,
 'booking_channel': 'App',
 'special_handling_flag': 0}

In [ ]:
db.customers.find_one()

{'_id': ObjectId('6a02e1fab9800d3fb76c0b7c'),
 'customer_id': 'C0001',
 'age': 26,
 'home_zone': 'North',
 'customer_type': 'SME',
 'signup_date': '2024-11-27 04:25:00',
 'loyalty_score': 44.9,
 'app_engagement_score': 69.2,
 'preferred_channel': 'App',
 'account_status': 'Active'}

**Converting DataFrames to MongoDB Documents and Insertion of the Documents**

A helper function was created to convert Pandas dataframes into JSON format suitable for MongoDB insertion. This ensures that structured tabular data is transformed into document-based format compatible with NoSQL storage.

In [7]:
def df_to_docs(df):
    return json.loads(df.to_json(orient='records', date_format='iso'))

for col in ["hubs","customers","drivers","vehicles",
            "orders","deliveries","incidents","complaints","app_events"]:
    db[col].drop()

db["hubs"].insert_many(df_to_docs(hubs))
db["customers"].insert_many(df_to_docs(customers))
db["drivers"].insert_many(df_to_docs(drivers))
db["vehicles"].insert_many(df_to_docs(vehicles))
db["orders"].insert_many(df_to_docs(orders))
db["deliveries"].insert_many(df_to_docs(deliveries))
db["incidents"].insert_many(df_to_docs(incidents))
db["complaints"].insert_many(df_to_docs(complaints))
db["app_events"].insert_many(df_to_docs(app_events))

print("\n✅ All collections inserted")
print("\nCollection counts:")
for col in sorted(db.list_collection_names()):
    print(f"  {col}: {db[col].count_documents({})} documents")


✅ All collections inserted

Collection counts:
  app_events: 640 documents
  complaints: 320 documents
  customers: 650 documents
  deliveries: 950 documents
  drivers: 170 documents
  hubs: 8 documents
  incidents: 280 documents
  orders: 1250 documents
  vehicles: 120 documents


###**MongoDB CRUD Operations**

**C - CREATE**

The Create operation in MongoDB was used to add new records into the collections. The insert_one() method was used to insert a single new driver record into the drivers collection, while insert_many() was used to add multiple complaint records at once into the complaints collection.

In [8]:
# CREATE - Insert new record into database
# Insert new driver's information
new_driver = {
    "driver_id" : "D171",
    "base_zone" : "West",
    "employment_type" : "FullTime",
    "years_experience" : 5,
    "training_score" : 60.2,
    "driver_rating" : 3.81,
    "shift_preference" : "Morning",
    "active_flag" : 1
}
insert_one_result = db.drivers.insert_one(new_driver)
print("Inserted new driver:", insert_one_result.inserted_id)

# Insert 2 complaints
new_complaints = [
    {
        "complaint_id": "CP0321",
        "customer_id": "C0081",
        "order_id": "O01111",
        "complaint_type": "Delay",
        "channel": "App",
        "severity": "High",
        "created_at": "2026-04-30T10:00:00",
        "status": "Open",
        "resolution_days": None,
        "compensation_amount": None
    },
    {
        "complaint_id": "CP0322",
        "customer_id": "C0087",
        "order_id": "O01113",
        "complaint_type": "MissedPickup",
        "channel": "Email",
        "severity": "Medium",
        "created_at": "2026-05-02T12:03:00",
        "status": "Resolved",
        "resolution_days": 5,
        "compensation_amount": 10.50
    }
]
insert_many_result = db.complaints.insert_many(new_complaints)
print("Inserted new complaints:", insert_many_result.inserted_ids)


Inserted new driver: 6a07224dfe6b4638e8cdfad9
Inserted new complaints: [ObjectId('6a07224dfe6b4638e8cdfada'), ObjectId('6a07224dfe6b4638e8cdfadb')]


**R - RETRIEVE**

The Retrieve function in MongoDB was utilized to obtain particular records from the database according to specified criteria. Queries were executed to filter late deliveries with lengthy route distances and to identify serious complaints submitted via the App channel. The find() method, combined with filtering criteria and projection fields, was utilized to obtain only the necessary information for analysis.

In [9]:
# RETRIEVE - Fetch data from database
# Find failed deliveries with long distance
delayed_long = db.deliveries.find(
    {"delivery_status": "Delayed", "route_distance_km" : {"$gt" : 15}},
    {"_id" : 0, "delivery_id" : 1, "order_id" : 1, "driver_id" : 1}
).limit(6)
for a in delayed_long:
    print(a)

# Find high severity complaints from App channel
high_chatbot = db.complaints.find(
    {"severity" : "High", "channel" : "App"},
    {"_id" : 0, "complaint_id" : 1, "complaint_type" : 1, "customer_id" : 1}
).limit(6)
for b in high_chatbot:
    print(b)


{'delivery_id': 'DL00004', 'order_id': 'O00313', 'driver_id': 'D116'}
{'delivery_id': 'DL00007', 'order_id': 'O00097', 'driver_id': 'D151'}
{'delivery_id': 'DL00017', 'order_id': 'O01249', 'driver_id': 'D002'}
{'delivery_id': 'DL00028', 'order_id': 'O00087', 'driver_id': 'D165'}
{'delivery_id': 'DL00052', 'order_id': 'O00554', 'driver_id': 'D168'}
{'delivery_id': 'DL00054', 'order_id': 'O01044', 'driver_id': 'D117'}
{'complaint_id': 'CP0001', 'customer_id': 'C0464', 'complaint_type': 'AppIssue'}
{'complaint_id': 'CP0063', 'customer_id': 'C0164', 'complaint_type': 'Billing'}
{'complaint_id': 'CP0096', 'customer_id': 'C0001', 'complaint_type': 'AppIssue'}
{'complaint_id': 'CP0112', 'customer_id': 'C0421', 'complaint_type': 'MissedPickup'}
{'complaint_id': 'CP0123', 'customer_id': 'C0361', 'complaint_type': 'DriverBehaviour'}
{'complaint_id': 'CP0133', 'customer_id': 'C0480', 'complaint_type': 'Delay'}


**U - UPDATE**

The Update operation in MongoDB was used to modify existing records within the collections. The update_one() method was applied to update the status and resolution details of a specific complaint, while update_many() was used to flag all vehicles currently under repair for priority maintenance. These operations helped maintain accurate and updated information within the database.

In [10]:
# UPDATE - Modify existing database records
# Update the complaint to resolved
update_one_result = db.complaints.update_one(
    {"complaint_id" : "CP0321"},
    {"$set" : {"status" : "Resolved", "resolution_days" : 2, "compensation_amount": 23.14}}
)
print("\nModified count:", update_one_result.modified_count)
updated = db.complaints.find_one({"complaint_id": "CP0321"}, {"_id": 0})
print("Updated document:", updated)

# Flag all the Inrepair vehicles
update_many_result = db.vehicles.update_many(
    {"maintenance_status": "InRepair"},
    {"$set": {"priority_maintenance": True}}
)
print(f"\nFlagged {update_many_result.modified_count} InRepair vehicles")


Modified count: 1
Updated document: {'complaint_id': 'CP0321', 'customer_id': 'C0081', 'order_id': 'O01111', 'complaint_type': 'Delay', 'channel': 'App', 'severity': 'High', 'created_at': '2026-04-30T10:00:00', 'status': 'Resolved', 'resolution_days': 2, 'compensation_amount': 23.14}

Flagged 36 InRepair vehicles


**D - DELETE**

The Delete operation in MongoDB was used to remove unnecessary records from the collections. The delete_one() method was applied to remove a specific driver record, while delete_many() was used to delete multiple complaint records simultaneously. These operations helped maintain accurate and organised database records.

In [11]:
# DELETE - Remove records from database
# Delete a driver's informaion
delete_one_result = db.drivers.delete_one(
    {"driver_id" : "D171"}
)
print("\nDeleted count:", delete_one_result.deleted_count)
print("Drivers remaining:", db.drivers.count_documents({}))

# Delete 2 complaints
delete_many_result = db.complaints.delete_many(
    {"complaint_id": {"$in": ["CP0321", "CP0322"]}}
)
print("\nDeleted count:", delete_many_result.deleted_count)
print("Complaints remaining:", db.complaints.count_documents({}))


Deleted count: 1
Drivers remaining: 170

Deleted count: 2
Complaints remaining: 320


###**MongoDB Aggregation Pipelines**

**Aggregation 1 - Delivery status summary**

This aggregation pipeline summarises deliveries based on their delivery status. It calculates the total number of deliveries, average customer ratings, and average fuel or charging costs for each delivery category such as OnTime, Delayed, and Failed. The results help evaluate delivery performance, customer satisfaction, and operational costs across different delivery outcomes.

In [12]:
# Delivery status summary
print("======================================================")
print("AGGREGATION 1: Delivery status summary")
print("======================================================")
pipeline1 = [
    {"$group": {
        "_id": "$delivery_status",
        "count": {"$sum": 1},
        "avg_rating": {"$avg": "$customer_rating_post_delivery"},
        "avg_fuel_cost": {"$avg": "$fuel_or_charge_cost"}
    }},
    {"$sort": {"count": -1}}
]
for doc in db["deliveries"].aggregate(pipeline1):
    print(f"  {doc['_id']:10} | count: {doc['count']:4} | "
          f"avg_rating: {doc['avg_rating']:.2f} | "
          f"avg_fuel: £{doc['avg_fuel_cost']:.2f}")

AGGREGATION 1: Delivery status summary
  OnTime     | count:  616 | avg_rating: 4.23 | avg_fuel: £12.68
  Delayed    | count:  202 | avg_rating: 3.04 | avg_fuel: £13.14
  Failed     | count:  132 | avg_rating: 3.03 | avg_fuel: £13.15


**Aggregation 2 - Complaints by type with average compensation**

This aggregation pipeline analyses complaints based on complaint type. It calculates the total number of complaints, average compensation amount, and average resolution days for each complaint category. The results help identify complaint types that occur more frequently and those that require higher compensation or longer resolution times.

In [13]:
# Complaints by type with avg compensation
print("==============================================================")
print("AGGREGATION 2: Complaints by type with average compensation")
print("==============================================================")
pipeline2 = [
    {"$group": {
        "_id": "$complaint_type",
        "total": {"$sum": 1},
        "avg_compensation": {"$avg": "$compensation_amount"},
        "avg_resolution_days": {"$avg": "$resolution_days"}
    }},
    {"$sort": {"total": -1}}
]
for doc in db["complaints"].aggregate(pipeline2):
    print(f"  {doc['_id']:20} | total: {doc['total']:3} | "
          f"avg_comp: £{doc['avg_compensation']:.2f} | "
          f"avg_days: {doc['avg_resolution_days']:.1f}")

AGGREGATION 2: Complaints by type with average compensation
  Delay                | total: 101 | avg_comp: £16.80 | avg_days: 7.3
  MissedPickup         | total:  64 | avg_comp: £22.24 | avg_days: 7.6
  AppIssue             | total:  53 | avg_comp: £18.50 | avg_days: 8.6
  DriverBehaviour      | total:  51 | avg_comp: £19.08 | avg_days: 8.2
  SupportExperience    | total:  20 | avg_comp: £17.12 | avg_days: 7.5
  Billing              | total:  16 | avg_comp: £23.87 | avg_days: 7.8
  Damage               | total:  15 | avg_comp: £23.98 | avg_days: 11.3


**Aggregation 3: Vehicles in repair by assigned zone**

This aggregation pipeline analyses vehicles currently under repair across different assigned zones. It calculates the total number of vehicles in repair and the average battery health percentage for each zone. The results help identify zones with higher maintenance issues and evaluate overall fleet condition.


In [14]:
# Vehicles in repair by assigned zone
print("========================================================")
print("AGGREGATION 3: Vehicles in repair by assigned zone")
print("========================================================")
pipeline3 = [
    {"$match": {"maintenance_status": "InRepair"}},
    {"$group": {
        "_id": "$assigned_zone",
        "count": {"$sum": 1},
        "avg_battery": {"$avg": "$battery_health_pct"}
    }},
    {"$sort": {"count": -1}}
]
for doc in db["vehicles"].aggregate(pipeline3):
    print(f"  Zone: {doc['_id']:10} | InRepair: {doc['count']} | "
          f"Avg Battery: {doc['avg_battery']:.1f}%")

AGGREGATION 3: Vehicles in repair by assigned zone
  Zone: RiverSide  | InRepair: 9 | Avg Battery: 77.5%
  Zone: Airport    | InRepair: 7 | Avg Battery: 71.9%
  Zone: North      | InRepair: 6 | Avg Battery: 78.1%
  Zone: East       | InRepair: 4 | Avg Battery: 78.9%
  Zone: Central    | InRepair: 4 | Avg Battery: 74.6%
  Zone: South      | InRepair: 3 | Avg Battery: 76.4%
  Zone: West       | InRepair: 3 | Avg Battery: 82.2%


 **Aggregation 4: Top drivers by failed deliveries**

 This aggregation pipeline identifies the top drivers with the highest number of failed deliveries. It calculates the total failed deliveries and the average number of manual route overrides for each driver. The results help evaluate driver performance and identify operational behaviours that may contribute to delivery failures.

In [15]:
# Top drivers by failed deliveries
print("======================================================")
print("AGGREGATION 4: Top drivers by failed deliveries")
print("======================================================")
pipeline4 = [
    {"$match": {"delivery_status": "Failed"}},
    {"$group": {
        "_id": "$driver_id",
        "failed_count": {"$sum": 1},
        "avg_overrides": {"$avg": "$manual_route_override_count"}
    }},
    {"$sort": {"failed_count": -1}},
    {"$limit": 5}
]
for doc in db["deliveries"].aggregate(pipeline4):
    print(f"  Driver: {doc['_id']} | Failed: {doc['failed_count']} | "
          f"Avg Overrides: {doc['avg_overrides']:.2f}")

AGGREGATION 4: Top drivers by failed deliveries
  Driver: D104 | Failed: 4 | Avg Overrides: 1.00
  Driver: D024 | Failed: 4 | Avg Overrides: 1.50
  Driver: D133 | Failed: 4 | Avg Overrides: 0.50
  Driver: D131 | Failed: 3 | Avg Overrides: 1.33
  Driver: D083 | Failed: 3 | Avg Overrides: 0.67


# **Query Optimisation Strategies**

**Importing Required Libraries**

The required Python modules were imported to support time tracking and MongoDB indexing operations. The time module was used for measuring query execution performance, while ASCENDING from PyMongo was imported to define index sorting order during index creation.

In [16]:
# Import
import time
from pymongo import ASCENDING

### **Index Creation**

**Before Indexing**

Before indexing, the query performance indicates that MongoDB needed to scan a large number of documents to retrieve results, leading to slower response times for certain operations. This highlights inefficiencies in data retrieval and shows the need for proper indexing to optimise query execution and improve overall system performance.

In [17]:
# Before Indexing
print("======================================================")
print("Before Indexing - Performance Test")
print("======================================================")

start = time.time()
list(db["deliveries"].find({"delivery_status": "Failed"}))
time1_before = (time.time()-start)*1000
print(f"\nQuery 1 (failed deliveries): {time1_before:.2f} ms")

start = time.time()
list(db["complaints"].find({"severity": "High", "status": "Open"}))
time2_before = (time.time()-start)*1000
print(f"Query 2 (high severity complaints): {time2_before:.2f} ms")

start = time.time()
list(db["vehicles"].find({"maintenance_status": "InRepair"}))
time3_before = (time.time()-start)*1000
print(f"Query 3 (vehicles in repair): {time3_before:.2f} ms")

start = time.time()
list(db["orders"].find({"pickup_zone": "Central"}))
time4_before = (time.time()-start)*1000
print(f"Query 4 (orders by zone): {time4_before:.2f} ms")

Before Indexing - Performance Test

Query 1 (failed deliveries): 247.87 ms
Query 2 (high severity complaints): 83.62 ms
Query 3 (vehicles in repair): 82.73 ms
Query 4 (orders by zone): 167.65 ms


**Before indexing plan** shows that MongoDB performs COLLSCAN (full collection scan) examining 950 documents and returned 132.

In [18]:
# Before Indexing Plan
print("======================================================")
print("Before Indexing Plan")
print("======================================================")
explain_before = db.command("explain",
    {"find": "deliveries", "filter": {"delivery_status": "Failed"}},
    verbosity="executionStats"
)
sb = explain_before['executionStats']
print(f"  Stage:         {explain_before['queryPlanner']['winningPlan']['stage']}")
print(f"  Docs Examined: {sb['totalDocsExamined']}")
print(f"  Docs Returned: {sb['nReturned']}")
print(f"  Execution ms:  {sb['executionTimeMillis']}")

Before Indexing Plan
  Stage:         COLLSCAN
  Docs Examined: 950
  Docs Returned: 132
  Execution ms:  0


**Creating Indexes**

In this step, indexes were created on key fields across multiple collections such as deliveries, complaints, vehicles, orders, and customers. These indexes help improve query performance by allowing MongoDB to locate records more efficiently instead of scanning entire collections. As a result, data retrieval becomes faster and more optimized for analytical queries.

In [19]:
# Creating Indexes
print("======================================================")
print("Creating Indexes")
print("======================================================")

for name in ["idx_delivery_status","idx_driver_id","idx_vehicle_id"]:
    try: db["deliveries"].drop_index(name)
    except: pass
for name in ["idx_severity_status","idx_customer_id"]:
    try: db["complaints"].drop_index(name)
    except: pass
try: db["vehicles"].drop_index("idx_maintenance_status")
except: pass
for name in ["idx_pickup_zone","idx_service_type", "idx_service_type_alt"]:
    try: db["orders"].drop_index(name)
    except: pass
try: db["customers"].drop_index("idx_home_zone")
except: pass

db["deliveries"].create_index([("delivery_status", ASCENDING)], name="idx_delivery_status")
db["deliveries"].create_index([("driver_id", ASCENDING)], name="idx_driver_id")
db["deliveries"].create_index([("vehicle_id", ASCENDING)], name="idx_vehicle_id")
print("✅ deliveries: 3 indexes created")

db["complaints"].create_index([("severity", ASCENDING), ("status", ASCENDING)], name="idx_severity_status")
db["complaints"].create_index([("customer_id", ASCENDING)], name="idx_customer_id")
print("✅ complaints: 2 indexes created")

db["vehicles"].create_index([("maintenance_status", ASCENDING)],
                              name="idx_maintenance_status")
print("✅ vehicles: 1 index created")

db["orders"].create_index([("pickup_zone", ASCENDING)], name="idx_pickup_zone")
db["orders"].create_index([("service_type", ASCENDING)], name="idx_service_type")
print("✅ orders: 2 indexes created")

db["customers"].create_index([("home_zone", ASCENDING)], name="idx_home_zone")
print("✅ customers: 1 index created")

Creating Indexes
✅ deliveries: 3 indexes created
✅ complaints: 2 indexes created
✅ vehicles: 1 index created
✅ orders: 2 indexes created
✅ customers: 1 index created


**After Indexing**

After applying indexing, query performance improved by reducing the time required to retrieve data across all operations. The system became more efficient in locating relevant documents, especially for frequently filtered fields such as delivery status and zone-based queries. Overall, indexing enhanced MongoDB query execution and improved responsiveness of the analytical operations.

In [20]:
# After Indexing
print("======================================================")
print("After Indexing - Performance Test")
print("======================================================")

start = time.time()
list(db["deliveries"].find({"delivery_status": "Failed"}))
time1_after = (time.time()-start)*1000
print(f"\nQuery 1 (failed deliveries): {time1_after:.2f} ms")

start = time.time()
list(db["complaints"].find({"severity": "High", "status": "Open"}))
time2_after = (time.time()-start)*1000
print(f"Query 2 (high severity complaints): {time2_after:.2f} ms")

start = time.time()
list(db["vehicles"].find({"maintenance_status": "InRepair"}))
time3_after = (time.time()-start)*1000
print(f"Query 3 (vehicles in repair): {time3_after:.2f} ms")

start = time.time()
list(db["orders"].find({"pickup_zone": "Central"}))
time4_after = (time.time()-start)*1000
print(f"Query 4 (orders by zone): {time4_after:.2f} ms")

After Indexing - Performance Test

Query 1 (failed deliveries): 246.44 ms
Query 2 (high severity complaints): 83.02 ms
Query 3 (vehicles in repair): 83.34 ms
Query 4 (orders by zone): 168.48 ms


**After indexing plan** shows that it uses FETCH to scan examining only 132 documents and returned 132.

In [21]:
# After Indexing Plan
print("======================================================")
print("After Indexing Plan")
print("======================================================")
explain_after = db.command("explain",
    {"find": "deliveries", "filter": {"delivery_status": "Failed"}},
    verbosity="executionStats"
)
sa = explain_after['executionStats']
print(f"  Stage:         {explain_after['queryPlanner']['winningPlan']['stage']}")
print(f"  Docs Examined: {sa['totalDocsExamined']}")
print(f"  Docs Returned: {sa['nReturned']}")
print(f"  Execution ms:  {sa['executionTimeMillis']}")

After Indexing Plan
  Stage:         FETCH
  Docs Examined: 132
  Docs Returned: 132
  Execution ms:  0


###**Performance analysis**

The performance summary compares query execution times before and after indexing to evaluate system efficiency improvements. Overall, the results show only minor changes in execution times, with some queries showing slight improvements while others remain largely unchanged due to minimal dataset scanning differences. This indicates that while indexing has a limited immediate impact on this dataset, it still contributes to more stable and consistent query performance.

In [22]:
# Performance Improvement Summary
print("======================================================")
print("Performance Summary")
print("======================================================")
print(f"\n  {'Query':<35} {'Before':>10} {'After':>10} {'Saving':>10}")
print(f"  {'-'*65}")
for label, b, a in [
    ("Failed deliveries",        time1_before, time1_after),
    ("High severity complaints", time2_before, time2_after),
    ("Vehicles in repair",       time3_before, time3_after),
    ("Orders by zone",           time4_before, time4_after),
]:
    saving = ((b-a)/b*100) if b > 0 else 0
    print(f"  {label:<35} {b:>9.2f}ms {a:>9.2f}ms {saving:>9.1f}%")

Performance Summary

  Query                                   Before      After     Saving
  -----------------------------------------------------------------
  Failed deliveries                      247.87ms    246.44ms       0.6%
  High severity complaints                83.62ms     83.02ms       0.7%
  Vehicles in repair                      82.73ms     83.34ms      -0.7%
  Orders by zone                         167.65ms    168.48ms      -0.5%
